File: 3524-find-x-value-of-array-i/3524-find-x-value-of-array-i.py
class Solution:
    def resultArray(self, nums: List[int], k: int) -> List[int]:
        n = len(nums)
        arr = [x % k for x in nums]
        res = [0] * k

        for req in range(k):
            dp = [[[-1] * k for _ in range(k + 1)] for _ in range(n)]

            def solve(i: int, prevProd: int) -> int:
                if i >= n:
                    return 0
                if dp[i][prevProd][req] != -1:
                    return dp[i][prevProd][req]

                skip = 0
                take = 0

                if prevProd == k:
                    skip = solve(i + 1, k)

                curProd = arr[i] if prevProd == k else (prevProd * arr[i]) % k
                if curProd == req:
                    take += 1
                take += solve(i + 1, curProd)

                total = skip + take
                dp[i][prevProd][req] = total
                return total

            res[req] = solve(0, k)

        return res

File: 3524-find-x-value-of-array-i/3524-find-x-value-of-array-i.rs
impl Solution {
    pub fn result_array(nums: Vec<i32>, k: i32) -> Vec<i64> {
        let n = nums.len();
        let k = k as usize;
        let arr: Vec<usize> = nums.iter().map(|&x| (x as usize) % k).collect();
        let mut res = vec![0i64; k];

        for req in 0..k {
            // dp[i][prevProd] = count
            let mut dp = vec![vec![-1i64; k + 1]; n];

            fn solve(
                i: usize,
                prev_prod: usize,
                req: usize,
                k: usize,
                arr: &Vec<usize>,
                dp: &mut Vec<Vec<i64>>,
            ) -> i64 {
                if i >= arr.len() {
                    return 0;
                }
                if dp[i][prev_prod] != -1 {
                    return dp[i][prev_prod];
                }

                let mut skip = 0i64;
                let mut take = 0i64;

                if prev_prod == k {
                    skip = solve(i + 1, k, req, k, arr, dp);
                }

                let cur_prod = if prev_prod == k {
                    arr[i]
                } else {
                    (prev_prod * arr[i]) % k
                };

                if cur_prod == req {
                    take += 1;
                }
                take += solve(i + 1, cur_prod, req, k, arr, dp);

                let total = skip + take;
                dp[i][prev_prod] = total;
                total
            }

            res[req] = solve(0, k, req, k, &arr, &mut dp);
        }

        res
    }
}